src/libsyntax_ext/deriving/cmp/partial_eq.rs

```diff
@@ -11,13 +11,33 @@
 use deriving::generic::*;
 use deriving::generic::ty::*;
 
-use syntax::ast::{MetaItem, Expr, BinOpKind};
+use syntax::ast::{MetaItem, Expr, BinOpKind, ItemKind, VariantData};
 use syntax::codemap::Span;
 use syntax::ext::base::{ExtCtxt, Annotatable};
 use syntax::ext::build::AstBuilder;
 use syntax::parse::token::InternedString;
 use syntax::ptr::P;
 
+fn is_clike_enum(item: &Annotatable) -> bool {
+    match *item {
+        Annotatable::Item(ref item) => {
+            match item.node {
+                ItemKind::Enum(ref enum_def, _) => {
+                    enum_def.variants.iter().all(|v|
+                        if let VariantData::Unit(..) = v.node.data {
```

Comment:
I mean, `enum E { V {} }` isn't considered "C-like", but it's still amenable to this `derive` optimization.